packages/Chem/src/rdkit-service/rdkit-service-worker-substructure.ts

```diff
@@ -174,4 +176,63 @@ export class RdKitServiceWorkerSubstructure extends RdKitServiceWorkerSimilarity
     console.log('Finished Worker ' + results.length);
     return results;
   }
+
+  getStructuralAlerts(alerts: {[rule in RuleId]?: string[]}, molecules?: string[]): {[rule in RuleId]?: boolean[]} {
+    if (this._rdKitMols === null && typeof molecules === 'undefined') {
+      console.debug(`getStructuralAlerts: No molecules to process`);
+      return {};
+    }
+
+    const ruleSmartsMap: {[rule in RuleId]?: (RDMol | null)[]} = {};
+    const rules = Object.keys(alerts) as RuleId[];
+    for (let rule of rules) {
+      const ruleLength = alerts[rule]!.length;
+      ruleSmartsMap[rule] = new Array(ruleLength);
+      for (let smartsIdx = 0; smartsIdx < ruleLength; smartsIdx++)
+        ruleSmartsMap[rule]![smartsIdx] = this.getQMol(alerts[rule]![smartsIdx]);
+    }
+
+    const molsCount = molecules?.length ?? this._rdKitMols!.length;
+    // Prepare the result storage
+    const resultValues: {[ruleId in RuleId]?: BitArray} = {};
+    for (const rule of rules)
+      resultValues[rule] = new BitArray(molsCount, false);
+
+    // Run the structural alerts detection
+    for (let molIdx = 0; molIdx < molsCount; molIdx++) {
+      const mol = typeof molecules !== 'undefined' ? getMolSafe(molecules[molIdx], {}, this._rdKitModule).mol :
+        this._rdKitMols![molIdx];
+      if (mol === null) {
+        console.debug(`Molecule ${molIdx} is null`);
+        continue;
+      }
+
+      for (const rule of rules) {
+        const ruleSmarts = ruleSmartsMap[rule]!;
+        for (let alertIdx = 0; alertIdx < ruleSmarts.length; alertIdx++) {
+          const smarts = ruleSmarts[alertIdx];
+          if (smarts === null) {
+            console.debug(`Smarts ${alertIdx} for rule ${rule} is null at molecule ${molIdx}`);
+            continue;
+          }
+
+          const matches = mol.get_substruct_match(smarts);
+          if (matches !== '{}') {
+            resultValues[rule]!.setTrue(molIdx);
+            break;
+          }
+        }
+      }
+      mol.delete();
+    }
+
+    for (const smartsList of Object.values(ruleSmartsMap)) {
+      for (const smarts of smartsList)
+        smarts?.delete();
+    }
+
+    this._rdKitMols = null;
+
+    return Object.fromEntries(Object.entries(resultValues).map(([k, val]) => [k, val.getRangeAsList(0, val.length)]));
+  }
 }
```

Comment:
What was the largest data set that you have tried and how many cores did you use? Was the time reduction in line with your expectations?

Reply:
Before it used to be about 8 seconds for all the alert groups on 1000 smiles. After the parallelization, it now takes only 4 seconds with 10 threads. The reason it's not less than a second now is that in each worker it takes time to convert smarts from structural alert dataset to RDKit molecules. I didn't test on larger data, but I assume the more data is being processed, the more run time reduction is.